core/src/Cuda/Kokkos_Cuda_Parallel_MDRange.hpp

```diff
@@ -95,11 +95,33 @@ class ParallelFor<FunctorType, Kokkos::MDRangePolicy<Traits...>, Kokkos::Cuda> {
 
   inline void execute() const {
     if (m_rp.m_num_tiles == 0) return;
-    const auto maxblocks = m_rp.space().cuda_device_prop().maxGridSize;
+    const auto maxblocks  = m_rp.space().cuda_device_prop().maxGridSize;
+    const auto maxthreads = m_rp.space().cuda_device_prop().maxThreadsDim;
+    [[maybe_unused]] const auto maxThreadsPerBlock =
+        m_rp.space().cuda_device_prop().maxThreadsPerBlock;
+    // make sure the Z dimension (it is less than x,y limits) isn't exceeded
+    const auto clampZ = [&](const int input) {
+      return (input > maxthreads[2] ? maxthreads[2] : input);
+    };
+    // make sure the block dimensions don't exceed the max number of threads
+    // allowed
+    const auto check_block_sizes = [&]([[maybe_unused]] const dim3& block) {
+      KOKKOS_ASSERT(block.x > 0 && block.x <= (unsigned int)maxthreads[0]);
+      KOKKOS_ASSERT(block.y > 0 && block.y <= (unsigned int)maxthreads[1]);
+      KOKKOS_ASSERT(block.z > 0 && block.z <= (unsigned int)maxthreads[2]);
+      KOKKOS_ASSERT(block.x * block.y * block.z <=
+                    (unsigned int)maxThreadsPerBlock);
```

Comment:
generally we prefer static_cast over c-cast, I will push a change to this branch, otherwise this is good.

Reply:
all good with me